imagetool/src/main/java/com/oracle/weblogic/imagetool/aru/AruPatch.java

```diff
@@ -0,0 +1,251 @@
+                AruPatch patch = new AruPatch()
+                    .patchId(XPathUtil.string(nodeList.item(i), "./name"))
+                    .version(XPathUtil.string(nodeList.item(i), "./release/@name"))
+                    .release(XPathUtil.string(nodeList.item(i), "./release/@id"))
+                    .releaseName(XPathUtil.string(nodeList.item(i), "./release/text()"))
+                    .description(XPathUtil.string(nodeList.item(i), "./bug/abstract"))
+                    .product(XPathUtil.string(nodeList.item(i), "./product/@id"))
+                    .psuBundle(XPathUtil.string(nodeList.item(i), "./psu_bundle"))
+                    .downloadHost(XPathUtil.string(nodeList.item(i), "./files/file/download_url/@host"))
+                    .downloadPath(XPathUtil.string(nodeList.item(i), "./files/file/download_url/text()"));
+
+                int index = patch.downloadPath().indexOf("patch_file=");
+                if (index < 0) {
+                    throw new XPathExpressionException(Utils.getMessage("IMG-0059", patch.patchId()));
+                }
+                patch.fileName(patch.downloadPath().substring(index + "patch_file=".length()));
+
+                logger.fine("AruPatch created id:" + patch.patchId()
+                    + "  ver:" + patch.version()
+                    + "  desc:" + patch.description()
+                    + "  rel:" + patch.release()
+                    + "  product:" + patch.product()
+                    + "  relName:" + patch.releaseName()
+                    + "  psu:" + patch.psuBundle()
+                    + "  url:" + patch.downloadUrl());
+                result.add(patch);
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Select a an ARU patch from the list based on a version number.
+     * Version preference is: provided version, PSU version, and then installer version.
+     * If there is only one patch in the list, no version checking is done, and that patch is returned.
+     * @param patches list of patches to search
+     * @param providedVersion user specified patch version, like bugnumber_12.2.1.4.0
+     * @param psuVersion version number of the PSU, if applicable
+     * @param installerVersion version of WebLogic installed or to be installed in the Oracle Home
+     * @return the selected patch, or null
+     * @throws VersionNotFoundException if the user requested a version that was not in the list.
+     */
+    public static AruPatch selectPatch(List<AruPatch> patches, String providedVersion, String psuVersion,
+                                       String installerVersion) throws VersionNotFoundException {
+
+        AruPatch selected = null;
+
+        if (patches.isEmpty()) {
+            return null;
+        }
+
+        if (patches.size() == 1) {
+            selected = patches.get(0);
+            if (selected.version() == null) {
+                if (providedVersion != null) {
```

Comment:
Is there any significance of providedVersion, psuVersion and installerVersion?  It uses providedVersion first before anything else?

Reply:
Yes, providedVersion is not null when the user requested a patch by version like 12341234_12.2.1.3.0.  psuVersion is not null when the user requested --latestPSU (or recommendedPatches), OR if this is UPDATE and the image already has a PSU applied.  installerVersion never null and matches the Oracle Home installer version (--version)